### PR TITLE
Created generic repository for sorting

### LINKED ISSUE #243: Make abstract class for custom sorting of groups and rooms

### PR DESCRIPTION
**Pull request by petrokrechunyak**

### CODE REVIEW

File: src/main/java/com/softserve/controller/GroupController.java [hunk 1]
```diff
@@ -97,15 +97,15 @@ public ResponseEntity<List<GroupDTO>> getDisabled() {
     @ApiOperation(value = "Create group ordered after another")
     public ResponseEntity<GroupDTO> saveGroupAfter(@RequestBody GroupOrderDTO groupDTO) {
```
Comment (dshma) [inline src/main/java/com/softserve/controller/GroupController.java]: method name is still "save"

File: src/main/java/com/softserve/controller/RoomController.java [hunk 2]
```diff
@@ -127,16 +127,19 @@ public ResponseEntity<List<RoomDTO>> getAllOrdered() {
     @PostMapping("/after/{id}")
     @ApiOperation(value = "Create room after id")
     public ResponseEntity<RoomDTO> saveRoomAfterId(@PathVariable("id") Long roomAfterId, @RequestBody RoomDTO roomDTO) {
+        if (roomAfterId == null) {
```
Comment (dshma) [inline src/main/java/com/softserve/controller/RoomController.java]: path variable are required by default
if it is missing, an exception should be thrown, how can this parameter be null considering that?

File: src/main/java/com/softserve/repository/SortOrderRepository.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import org.springframework.stereotype.Repository;
+
+import javax.persistence.TypedQuery;
+import java.util.Optional;
+
+@Slf4j
+@Repository
+@Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class SortOrderRepository<T extends SortableOrder> {
+    public SortOrderRepository() {
```
Comment (dshma) [inline src/main/java/com/softserve/repository/SortOrderRepository.java]: why is this constructor needed?

File: src/main/java/com/softserve/repository/SortOrderRepository.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        }
+        sessionFactory.getCurrentSession().update(t);
+        return t;
+    }
+
+    public boolean isExistsById(Long id) {
+        log.info("In isExistsById(id = [{}])", id);
+        return (boolean) sessionFactory.getCurrentSession()
+                .createQuery(
+                        "SELECT (count(*) > 0)"
```
Comment (dshma) [inline src/main/java/com/softserve/repository/SortOrderRepository.java]: SELECT 1 or SELECT EXISTS could be used, I'd recommend the first approach, in that case uniqueResult() != null can be checked

File: src/main/java/com/softserve/repository/SortOrderRepository.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                        + "WHERE c.id = :id")
+                .setParameter("id", id)
+                .getSingleResult();
+    }
+
+    public void changeOrderOffset(Integer lower, Integer upper) {
+        log.info("Entered changeOrderOffset({}, {})", lower, upper);
+        TypedQuery<?> typedQuery = sessionFactory.getCurrentSession().createQuery(
+                "UPDATE " + tClass.getSimpleName() + " c "
+                + "SET c.sortOrder = c.sortOrder+1 "
```
Comment (dshma) [inline src/main/java/com/softserve/repository/SortOrderRepository.java]: missing spaces

File: src/main/java/com/softserve/repository/SortOrderRepository.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Optional;
+
+@Slf4j
+@Repository
+@Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class SortOrderRepository<T extends SortableOrder> {
+    public SortOrderRepository() {
+    }
+
+    private Class<?> tClass = SortOrderRepository.class;
```
Comment (dshma) [inline src/main/java/com/softserve/repository/SortOrderRepository.java]: why initialize it like this?

Comment (dshma) [inline src/main/java/com/softserve/repository/SortOrderRepository.java]: the setter is a bid bizarre in this case, let's follow this naming convention:
https://stackoverflow.com/questions/2529974/why-do-java-programmers-like-to-name-a-variable-clazz

**Review by dshma — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserve/repository/SortOrderRepository.java (hunk 5 above)
Comment (petrokrechunyak) [reply src/main/java/com/softserve/repository/SortOrderRepository.java]: I can`t understand which spaces did I miss.

Comment (dshma) [reply src/main/java/com/softserve/repository/SortOrderRepository.java]: c.sortOrder+1 vs. c.sortOrder + 1

**Review by dshma — APPROVED** [review]

**Pull request merged**